### PR TITLE
Added support for CKM_RSA_PKCS_PSS

### PR DESCRIPTION
**Pull request by nmav**
This mechanism is the most likely to be used by higher level
software utilizing RSA-PSS as it is in par with CKM_RSA_PKCS
and does not require to provide the whole data to be hashed.

Signed-off-by: Nikos Mavrogiannopoulos <nmav@redhat.com>

### CODE REVIEW

**Comment by nmav** [comment]
I'd appreciate a thorough review, as well as some help with the test suite. I'm unable to successfully run it on develop branch (with or without my changes).

**Comment by bellgrim** [comment]
Thank you for this pull request, I will review it within a couple of weeks.

File: src/lib/SoftHSM.cpp [hunk 1]
```diff
@@ -670,6 +670,9 @@ CK_RV SoftHSM::C_GetMechanismList(CK_SLOT_ID slotID, CK_MECHANISM_TYPE_PTR pMech
 		CKM_SHA256_RSA_PKCS,
 		CKM_SHA384_RSA_PKCS,
 		CKM_SHA512_RSA_PKCS,
+#ifdef WITH_OPENSSL
```
Comment (bellgrim) [inline src/lib/SoftHSM.cpp:677]: You also need to increase the number of supported mechanisms above. E.g:

```
#ifdef WITH_OPENSSL
 	nrSupportedMechanisms += 1;
#endif
```

File: src/lib/test/SignVerifyTests.cpp [hunk 2]
```diff
@@ -195,6 +196,44 @@ void SignVerifyTests::signVerifySingle(CK_MECHANISM_TYPE mechanismType, CK_SESSI
...
+void SignVerifyTests::signVerifySingleData(size_t dataSize, CK_MECHANISM_TYPE mechanismType, CK_SESSION_HANDLE hSession, CK_OBJECT_HANDLE hPublicKey, CK_OBJECT_HANDLE hPrivateKey, CK_VOID_PTR param /* = NULL_PTR */, CK_ULONG paramLen /* = 0 */)
+{
+	CK_RV rv;
+	CK_MECHANISM mechanism = { mechanismType, param, paramLen };
+	CK_BYTE *data = (CK_BYTE*)malloc(dataSize);
+	CK_BYTE signature[1024];
+	CK_ULONG ulSignatureLen = 0;
+	unsigned i;
+
+	assert(data != NULL);
```
Comment (bellgrim) [inline src/lib/test/SignVerifyTests.cpp]: Change to CPPUNIT_ASSERT so it is handled by the test framework

```
CPPUNIT_ASSERT(data != NULL);
```

File: src/lib/test/SignVerifyTests.cpp [hunk 3]
```diff
@@ -42,6 +42,7 @@
 #include <config.h>
 #include <stdlib.h>
 #include <string.h>
+#include <assert.h>
```
Comment (bellgrim) [inline src/lib/test/SignVerifyTests.cpp]: Not needed if we change to CPPUNIT_ASSERT as noted in the next comment.

File: src/lib/crypto/OSSLRSA.cpp [hunk 4]
```diff
@@ -121,6 +121,104 @@ bool OSSLRSA::sign(PrivateKey* privateKey, const ByteString& dataToSign,
...
+
+		RSA* rsa = osslKey->getOSSLKey();
+
+		if (dataToSign.size() != allowedLen)
+		{
+			ERROR_MSG("Data to sign does not match expected (%d) for RSA PSS", (int)allowedLen);
+
+			return false;
+		}
+
```
Comment (bellgrim) [inline src/lib/crypto/OSSLRSA.cpp:184]: Should also check the salt length:

```
size_t sLen = pssParam->sLen;
if (sLen > ((privateKey->getBitLength()+6)/8-2-allowedLen))
{
    ERROR_MSG("sLen (%lu) is too large for current key size (%lu)",
    (unsigned long)sLen, privateKey->getBitLength());
    return false;
}

File: src/lib/crypto/OSSLRSA.cpp [hunk 5]
```diff
@@ -607,6 +705,82 @@ bool OSSLRSA::verify(PublicKey* publicKey, const ByteString& originalData,
...
+			allowedLen = 64;
+			break;
+		default:
+			return false;
+		}
+
+		if (originalData.size() != allowedLen) {
+			return false;
+		}
+
```
Comment (bellgrim) [inline src/lib/crypto/OSSLRSA.cpp:787]: And also check the salt length here.

File: src/lib/SoftHSM.cpp (hunk 1 above)
Comment (nmav) [reply src/lib/SoftHSM.cpp:677]: Thanks. Included in the updated patch.

File: src/lib/test/SignVerifyTests.cpp (hunk 3 above)
Comment (nmav) [reply src/lib/test/SignVerifyTests.cpp]: Included in the updated patch.

File: src/lib/crypto/OSSLRSA.cpp (hunk 4 above)
Comment (nmav) [reply src/lib/crypto/OSSLRSA.cpp:184]: Included in the updated patch.

File: src/lib/crypto/OSSLRSA.cpp (hunk 5 above)
Comment (nmav) [reply src/lib/crypto/OSSLRSA.cpp:787]: Included in the updated patch.

**Comment by bellgrim** [comment]
Thank you for fixing this, will merge this code. Botan did not have support for PSS signing of raw hashes, but it has now been added. I will add some code for that.

**Pull request merged**